File: book_shop.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    // Fast I/O (optional, but recommended for processing large inputs)
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int num_books, budget;
    if (!(cin >> num_books >> budget)) return 0;

    vector<int> books_prices(num_books);
    for (int i = 0; i < num_books; i++) {
        cin >> books_prices[i];
    }

    vector<int> pages_per_book(num_books);
    for (int i = 0; i < num_books; i++) {
        cin >> pages_per_book[i];
    }

    // Initialize DP table. Size is (num_books + 1) x (budget + 1)
    // We pad an extra row at the top to act as the base case (zero books considered)
    vector<vector<int>> dp(num_books + 1, vector<int>(budget + 1, 0));

    for (int i = 1; i <= num_books; i++) {
        for (int j = 0; j <= budget; j++) {
            
            // i is 1-indexed in the DP table, but our arrays are 0-indexed.
            // Therefore, book[i-1] represents the current book being evaluated.
            int skip = dp[i-1][j];
            int take = 0;
            
            if (j - books_prices[i-1] >= 0) {
                take = dp[i-1][j - books_prices[i-1]] + pages_per_book[i-1];
            }
            
            dp[i][j] = max(skip, take);
        }
    }

    // The final answer is in the bottom-right corner of the DP matrix
    cout << dp[num_books][budget] << "\n";

    return 0;
}

File: book_shop.py

from functools import cache




num_books, budget = map(int, input().split())

books_prices = list(map(int, input().split()))

pages_per_book = list(map(int, input().split()))

dp = [[0 for i in range(budget + 1)] for j in range(num_books)]



for i in range(num_books):
    for j in range(budget + 1):
        skip = dp[i-1][j]
        take = 0
        if j - books_prices[i] >= 0:
            take = dp[i-1][j-books_prices[i]] + pages_per_book[i]
        dp[i][j] = max(skip, take)
print(dp[-1][-1])


@cache
def dfs(budget, i):
    if i >= num_books or budget < 0:
        return 0
    
    skip = 0 + dfs(budget, i + 1)
    take = 0
    if budget - books_prices[i] >= 0:
        take = pages_per_book[i] + dfs(budget - books_prices[i], i + 1)

    return max(skip, take)
